File: medium/_1769_MinimumNumberofOperations.cpp
// 1769. Minimum Number of Operations to Move All Balls to Each Box    

// https://leetcode.com/problems/minimum-number-of-operations-to-move-all-balls-to-each-box/discuss/1075669/C%2B%2B-O(N)-keep-track-of-of-balls-to-the-right-and-left

class Solution {
public:
    vector<int> minOperations(string boxes) {
        int N = boxes.size();  
        vector<int> output(N);        
                
        int sum = 0;
        int right = 0;
        int left = 0;
        for (int i = 0; i < N; ++i) 
        {
            if (boxes[i] == '1') 
            {
                sum += i;
                ++right;
            }
        }
        
        for (int i = 0; i < N; ++i) 
        {
            output[i] = sum;
            if (boxes[i] == '1') 
            {
                --right;
                    ++left;
            }
            sum += left - right;
        }
        return output;
    }
};

File: medium/_1769_MinimumNumberofOperations.py
# 1769. Minimum Number of Operations to Move All Balls to Each Box    

"""https://leetcode.com/problems/minimum-number-of-operations-to-move-all-balls-to-each-box/discuss/1075518/Clean-Python-3-two-passes-O(N)"""

"""看不懂"""

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        n = len(boxes)
        output, num_right, num_left, cnt_right, cnt_left = [0] * n, 0, 0, 0, 0
        
        for i in range(n):

            output[i] += cnt_right * i 
            output[i] -= num_right
            cnt_right += int(boxes[i])
            num_right += i * int(boxes[i])

            output[n - i - 1] += cnt_left * i
            output[n - i - 1] -= num_left
            cnt_left += int(boxes[n - i - 1])
            num_left += i * int(boxes[n - i - 1])

        return output